internal/app/server/impl.go

```diff
@@ -0,0 +1,374 @@
+			resp := &v1alpha1.BidiStreamResponse{
+				Payload: payload,
+			}
+			err := sendOnStream[v1alpha1.BidiStreamResponse](
+				ctx,
+				stream,
+				resp,
+				responseDefinition.ResponseDelayMs,
+			)
+			if err != nil {
+				return err
+			}
+
+		}
+	}
+
+	if responseDefinition.Error != nil {
+		return createError(responseDefinition.Error)
+	}
+	return nil
+}
+
+// NewConformanceServiceHandler returns a new ConformanceServiceHandler.
+func NewConformanceServiceHandler() conformancev1alpha1connect.ConformanceServiceHandler {
+	return &conformanceServer{}
+}
+
+// Sends a response T on the given stream. If delayMs is > 0, the function
+// will wait that many milliseconds before sending the message.
+func sendOnStream[T any](
+	ctx context.Context,
+	stream Stream[T],
+	resp *T,
+	delayMs uint32,
+) error {
+	var ticker *time.Ticker
+	if delayMs > 0 {
+		ticker = time.NewTicker(time.Duration(delayMs) * time.Millisecond)
+		defer ticker.Stop()
+	}
+	if ticker != nil {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+		}
+	}
+
+	if err := stream.Send(resp); err != nil {
+		return fmt.Errorf("error sending on stream: %w", err)
+	}
+	return nil
+}
+
+// Parses the given unary response definition and returns either
+// a built payload or a connect error based on the definition.
+func parseUnaryResponseDefinition(
+	def *v1alpha1.UnaryResponseDefinition,
+	headers http.Header,
+	reqs []*anypb.Any,
+) (*v1alpha1.ConformancePayload, error) {
+	if def != nil {
+		switch rt := def.Response.(type) {
+		case *v1alpha1.UnaryResponseDefinition_Error:
+			return nil, createError(rt.Error)
+		case *v1alpha1.UnaryResponseDefinition_ResponseData, nil:
+			payload := initPayload(headers, reqs)
+
+			// If response data was provided, set that in the payload response
+			if rt, ok := rt.(*v1alpha1.UnaryResponseDefinition_ResponseData); ok {
+				payload.Data = rt.ResponseData
+			}
+			return payload, nil
+		default:
+			return nil, connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("provided UnaryRequest.Response has an unexpected type %T", rt))
+		}
+	}
+	return nil, connect.NewError(connect.CodeInvalidArgument, errors.New("no response definition provided"))
+}
+
+// Initializes a conformance payload
+func initPayload(headers http.Header, reqs []*anypb.Any) *v1alpha1.ConformancePayload {
+	payload := &v1alpha1.ConformancePayload{}
+
+	headerInfo := []*v1alpha1.Header{}
+	for key, value := range headers {
+		hdr := &v1alpha1.Header{
+			Name:  key,
+			Value: value,
+		}
+		headerInfo = append(headerInfo, hdr)
+	}
+
+	// Set all observed request headers and requests in the response payload
+	payload.RequestInfo = &v1alpha1.ConformancePayload_RequestInfo{
+		RequestHeaders: headerInfo,
+		Requests:       reqs,
+	}
+	return payload
+}
+
+// Sets all response headers and trailers onto the given response
+func setResponseHeadersAndTrailers[T any](def ConformanceRequest, resp *connect.Response[T]) {
```

Comment:
I think we can simplify this -- get rid of the generics and make it work with unary and streaming responses -- by just making this a simple helper to translate `[]v1alpha1.Header` -> `http.Header`.
```suggestion
// Adds all header values in src to dest.
func addHeaders(
    src []*v1alpha1.Header,
    dest http.Header,
) {
```
The caller then does:
```go
addHeaders(respDefinition.ResponseHeaders, resp.Header())
addHeaders(respDefinition.ResponseTrailers, resp.Trailer())
```
And then you don't need the duplicated logic for the stream case.